### ISSUE TITLE
Consider not having style for disjoint layout options in the same struct

### ISSUE THREAD

**Issue by stephanemagnenat** [DioxusLabs/taffy#605]
## What problem does this solve or what need does it fill?

At the moment, the `Style` struct contains both properties common to all layout options (e.g., `margin`), but also specific ones to `Flexbox`, `Grid`, etc. This feels wrong to me, as it consumes extra memory that is not needed (if I'm not mistaken, the `Flexbox` properties do not make sense when using a `Grid` layout). 

## What solution would you like?

The layout-specific properties could be stored inside inner structs specific to each variant of the `Display` enum. This would possibly allow to keep each of these inner structs in their own modules, reducing the amount of `#[cfg(feature = "...")]" needed inside the style top module.

## What alternative(s) have you considered?

The current approach can be kept, it doesn't hurt, it just uses a bit of extra memory and makes the `Style` struct quite big.

## Additional context

I am considering using `taffy` for my own GUI framework, to be open sourced one day if/when it exists enough.

**Comment by Patryk27** [comment]
Note that a (huge, in my opinion) downside of using an `enum` would be that changing the display mode would remove current values from the memory.

That is, patterns such as (in pseudo-code):

```rust
fn initialize(ui: &mut Ui) {
    widget.display = Display::Flex;
    widget.align_self = AlignSelf::FlexEnd;
    widget.flex_direction = FlexDirection::RowReverse;
}

fn update(ui: &mut Ui) {
    if something {
        widget.display = Display::Flex;
    } else {
        widget.display = Display::None;
    }
}
```

... would stop working, as usesr would need to re-do the initialization once again during the update; that's not the worst thing in the world, of course, but it's very handy.

**Comment by stephanemagnenat** [comment]
That is true that these use cases would get harder. However, I am wondering whether such patterns are not side effects of the web/CSS habits rather than good programming practices. Because in principle, if there is a specific `FlexInner` configuration to use, I would expect that this specific configuration could just be a constant somewhere:
```rust
const FLEX_CONFIG: FlexInner = FlexInner {
     align_self: AlignSelf::FlexEnd,
     flex_direction = FlexDirection::RowReverse,
    ...
};

fn initialize(ui: &mut Ui) {
    widget.display = Display::Flex(FLEX_CONFIG);
}

fn update(ui: &mut Ui) {
    if something {
        widget.display = Display::Flex(FLEX_CONFIG);
    } else {
        widget.display = Display::None;
    }
}
```
(assuming `FlexInner` is `Copy`)

**Comment by nicoburns** [comment]
There's a bit of a tradeoff between memory usage and layout performance here (as any compressed representation will need to be translated when performing layout (which is currently made more significant by the fact the algorithm code will request access to the style multiple times during the layout of each node).

I think the plan is something like https://github.com/DioxusLabs/taffy/pull/568 which will allow consumers of Taffy to bring their own style representation (although the details of the implementation may well change). That could be:

- A great big struct like Taffy currently uses
- A smaller struct supporting only a subset of the properties
- Something like a `Vec<Property>`
- Style properties that are Boxed/Arc'ed/Arena allocated in some way
- etc
